plotter/contour.go

```diff
@@ -0,0 +1,494 @@
+	// by Min and Max.
+	Underflow color.Color
+	Overflow  color.Color
+
+	// Min and Max define the dynamic range of the
+	// heat map.
+	Min, Max float64
+}
+
+// NewContour creates as new contour plotter for the given data,
+// using the provided palette. If g has Min and Max methods that return
+// a float, those returned values are used to set the respective HeatMap
+// fields.
+func NewContour(g GridXYZ, levels []float64, p palette.Palette) *Contour {
+	var min, max float64
+	type minMaxer interface {
+		Min() float64
+		Max() float64
+	}
+	switch g := g.(type) {
+	case minMaxer:
+		min, max = g.Min(), g.Max()
+	default:
+		min, max = math.Inf(1), math.Inf(-1)
+		c, r := g.Dims()
+		for i := 0; i < c; i++ {
+			for j := 0; j < r; j++ {
+				v := g.Z(i, j)
+				if math.IsNaN(v) {
+					continue
+				}
+				min = math.Min(min, v)
+				max = math.Max(max, v)
+			}
+		}
+	}
+
+	if len(levels) == 0 {
+		levels = quantilesR7(g, defaultQuantiles)
+	}
+
+	return &Contour{
+		GridXYZ:   g,
+		Levels:    levels,
+		LineStyle: DefaultLineStyle,
+		Palette:   p,
+		Min:       min,
+		Max:       max,
+	}
+}
+
+// Default quantiles for case where levels is not explicitly set.
+var defaultQuantiles = []float64{0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99}
+
+// quantilesR7 returns the pth quantiles of the data in g according the th R-7 method.
+// http://en.wikipedia.org/wiki/Quantile#Estimating_the_quantiles_of_a_population
+func quantilesR7(g GridXYZ, p []float64) []float64 {
+	c, r := g.Dims()
+	data := make([]float64, 0, c*r)
+	for i := 0; i < c; i++ {
+		for j := 0; j < r; j++ {
+			if v := g.Z(i, j); !math.IsNaN(v) {
+				data = append(data, v)
+			}
+		}
+	}
+	sort.Float64s(data)
+	v := make([]float64, len(p))
+	for j, q := range p {
+		if q == 1 {
+			v[j] = data[len(data)-1]
+		}
+		h := float64(len(data)-1) * q
+		i := int(h)
+		v[j] = data[i] + (h-math.Floor(h))*(data[i+1]-data[i])
+	}
+	return v
+}
+
+// naive is a debugging constant. If true, Plot performs no contour path
+// reconstruction, instead rendering wach path segment individually.
+const naive = false
+
+// Plot implements the Plot method of the plot.Plotter interface.
+func (h *Contour) Plot(c draw.Canvas, plt *plot.Plot) {
+	var pal []color.Color
+	if h.Palette != nil {
+		pal = h.Palette.Colors()
+	}
+	if len(pal) == 0 {
+		pal = []color.Color{h.LineStyle.Color}
+	}
+
+	c.SetLineStyle(h.LineStyle)
+
+	trX, trY := plt.Transforms(&c)
+
+	// Collate contour paths and draw them.
+	//
+	// The alternative naive approach is to draw each line segment as
+	// conrec returns it. The integrated path approach allows graphical
+	// optimisations and is necessary for contour fill shading.
+	if !naive {
+		cp := contourPaths(h.GridXYZ, h.Levels, trX, trY)
+		ps := float64(len(pal)-1) / (h.Levels[len(h.Levels)-1] - h.Levels[0])
+		if len(h.Levels) == 1 {
+			ps = 0
+		}
+
+		for _, z := range h.Levels {
+			if math.IsNaN(z) {
+				continue
+			}
+			for _, pa := range cp[z] {
+				if isLoop(pa) {
```

Comment:
Sorry, I don't understand this (probably due to not really understanding vg.Path). If the path starts and ends at the same place, why does it need to have a line segment attaching the point to itself?